src/Renci.SshNet/Common/Pipe.cs

```diff
@@ -0,0 +1,47 @@
+﻿namespace Renci.SshNet.Common
+{
+    using System;
+    using System.IO;
+
+    /// <summary>
+    /// A generic pipe to pass through data.
+    /// </summary>
+    internal class Pipe : IDisposable
+    {
```

Comment:
I indeed agree that we should only block until at least some data is available or the writer has signaled EOS.

Was offline for a few hours to install outdoor speakers. Why do this in the summer if you can do it while it raining and freezing? :p

The **PipeOutputStream** can share the queue with the **PipeInputStream**.
The details can be discussed as we proceed.
Since we only expose **Stream** publicly, we can easily modify our internal design.